webapp/app/routes/app/emergency/$id/_route.tsx

```diff
@@ -0,0 +1,259 @@
+      emergencyBoard: boardAddress,
+      zkFoundation: await zkFoundationAddress(),
+    },
+    user,
+    signatures,
+    allGuardians: await guardianMembers(),
+    allSecurityCouncil: await councilMembers(),
+  });
+}
+
+function extract<T extends ZodTypeAny>(
+  formData: FormData,
+  key: string,
+  parser: T
+): z.infer<typeof parser> {
+  const value = formData.get(key);
+  const parsed = parser.safeParse(value);
+  if (parsed.error) {
+    throw badRequest(`Wrong value for ${key}`);
+  }
+  return parsed.data;
+}
+const intentParser = z.enum(["newSignature", "broadcastSuccess"]);
+
+export async function action({ request }: ActionFunctionArgs) {
+  const formData = await request.formData().catch(() => {
+    throw badRequest("Failed to parse body");
+  });
+
+  const intent = extract(formData, "intent", intentParser);
+  const proposalId = extract(formData, "proposalId", zodHex);
+
+  if (intent === intentParser.enum.newSignature) {
+    const user = requireUserFromHeader(request);
+    const signature = extract(formData, "signature", zodHex);
+    const actionName = extract(formData, "actionName", actionSchema);
+
+    await saveEmergencySignature(signature, user.address, actionName, proposalId);
+  }
+
+  if (intent === intentParser.enum.broadcastSuccess) {
+    await broadcastSuccess(proposalId);
+  }
+
+  return json({ ok: true });
+}
+
+const ActionSchema = z.enum([
+  "ExecuteEmergencyUpgradeGuardians",
+  "ExecuteEmergencyUpgradeSecurityCouncil",
+  "ExecuteEmergencyUpgradeZKFoundation",
+]);
+type Action = z.infer<typeof ActionSchema>;
+
+const ACTION_NAMES = {
+  guardian: ActionSchema.enum.ExecuteEmergencyUpgradeGuardians,
+  securityCouncil: ActionSchema.enum.ExecuteEmergencyUpgradeSecurityCouncil,
+  zkFoundation: ActionSchema.enum.ExecuteEmergencyUpgradeZKFoundation,
+};
+
+function actionForRole(role: UserRole): Action {
+  if (role === UserRoleSchema.enum.visitor) {
+    throw new Error("Visitors are not allowed to sign emergency upgrades");
+  }
+
+  return ACTION_NAMES[role];
+}
+
+export default function EmergencyUpgradeDetails() {
+  const navigate = useNavigate();
+
+  const { user, proposal, addresses, signatures, allSecurityCouncil, allGuardians } =
+    useLoaderData<typeof loader>();
+
+  if (user.role === "visitor") {
+    return "Unauthorized: Only valid signers can see this page.";
+  }
+
+  const actionName = actionForRole(user.role);
+  const haveAlreadySigned = signatures.some((s) => isAddressEqual(s.signer, user.address as Hex));
+  const gatheredScSignatures = signatures.filter((sig) => {
+    return allSecurityCouncil.some((addr) => isAddressEqual(addr, sig.signer));
+  }).length;
+  const gatheredGuardianSignatures = signatures.filter((sig) => {
+    return allGuardians.some((addr) => isAddressEqual(addr, sig.signer));
+  }).length;
+  const gatheredZkFoundationSignatures = signatures.filter((s) =>
+    isAddressEqual(s.signer, addresses.zkFoundation)
+  ).length;
+  return (
+    <>
```

Comment:
This view is being pushed to the bottom so it looks weird:
![image](https://github.com/user-attachments/assets/f1d40726-b585-4e11-af64-975b43bf26b4)

---

Consider sorting out the divs here like:

```tsx
  return (
    <div className="flex flex-col min-h-screen">
      <div className="mb-4 flex items-center pl-2">
        <Button
          size="icon"
          variant="ghost"
          onClick={() => navigate(-1)}
          className="mr-2 hover:bg-transparent"
        >
          <ArrowLeft />
        </Button>
        <h2 className="font-semibold">Proposal {proposal.externalId}</h2>
      </div>
      <div className="grid grid-cols-1 gap-4 md:grid-cols-2 mb-8">
...
```